### DISCUSSION TITLE
Minimize UDP packet size

### DISCUSSION THREAD

**Discussion by Gerbaum** [bluenviron/mediamtx#317] [Questions]
## Which version are you using?

v0.13.2

Ubuntu Arm6

## Question

I stream a USB Camera with following pipe to rtsp-simple-server:

gst-launch-1.0 rtspclientsink name=s location=rtsp://localhost:8554/mystream v4l2src device=/dev/v4l/by-id/usb-5M_Camera_5M_Camera_5M_Camera-video-index0 ! image/jpeg,width=1280,height=720,framerate=30/1 ! videorate drop-only=true max-rate=10 ! image/jpeg,framerate=10/1 ! jpegdec ! queue ! videoconvert ! queue !  v4l2h264enc extra-controls="controls,h264_profile=4,h264_level=10,video_bitrate=1000000;" !  s.sink_1

Simple rtsp server runs with default config.

I have set up an IPsec Ikev2 connection to a remote station, where I get the stream with VLC Player.

Wireshark tells me, that on the player side my packets are fragmented, so I reduced interface MTU value on server side.

But that showed no effect, I guess that simple-server reduces UDP packets according to interface MTU, but that is never enough to get a not fragmented packet size.

Any idea how I can reduce UDP packet size so it fits i one packet with Ipsec header?

**Comment by aler9** [comment]
Hello, packets are generated by GStreamer, not by rtsp-simple-server; you can tune the UDP packet size by using the `rtp-blocksize` parameter in `rtspclientsink`:

```
gst-launch-1.0 rtspclientsink name=s rtp-blocksize=500 location=rtsp://localhost:8554/mystream
```

Where 500 is the desired payload size. To get the resulting packet size, you have to do
500 +
UDP header size (8) +
IPsec header size (64) =
572

**Comment by saket424** [comment]
@aler9 
Any equivalent command to set rtp payload size in ffmpeg?

ffmpeg -re -stream_loop -1 -i file.ts -c copy -f rtsp rtsp://localhost:8554/mystream?pkt_size=1316 does not appear to be heeded

**Reply by tmatth** [reply]
You should double check that you're pushing UDP via `ffmpeg`, the pkt_size option was ignored for TCP until very recently and AFAIK is not available in a release yet, see:
http://ffmpeg.org/pipermail/ffmpeg-devel/2022-April/295698.html

**Comment by saket424** [comment]
I am not attempting rtsp_transport tcp at all. I want the udp rtsp to heed the pkt_size

ffmpeg -re -stream_loop -1 -i file.ts -c copy -f rtsp rtsp://localhost:8554/mystream?pkt_size=1316

whereas this works

ffmpeg -re -stream_loop -1 -i file.ts -c copy -f rtp rtp://localhost:1234?pkt_size=1316

**Reply by tmatth** [reply]
Yeah that was my understanding but IIRC ffmpeg may fall back to TCP transport if there's an issue with UDP, in which case `pkt_size` won't be respected. Even with UDP, you also need a fairly recent ffmpeg for that option to work FWIW.

**Reply by tmatth** [reply]
Oh also I just noticed that you're using the URL query parameter, try instead with `-pkt_size 1316`

**Reply by saket424** [reply]
@tmatth 
Adding -pkt_size 1316 did not seem to help with the old version of ffmpeg I have. The symptom i am seeing is the feed is not playable over t-mobile cellular but playable over wifi -- i am suspecting it has something to do with MTU

**Reply by tmatth** [reply]
Yeah you need ffmpeg 4.2 or later.

**Reply by saket424** [reply]
@tmatth 
I tried a new ffmpeg 5.0.1 statically build from here https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz

No difference. cellular still does not work

**Reply by tmatth** [reply]
You've tried smaller sizes like @aler9 suggested?

**Reply by saket424** [reply]
@tmatth 

I can confirm via tcpdump that the -pkt_size 1316 option is indeed working

21:51:05.538147 IP xyz.com.8104 > abc.com.8000: UDP, length 1316
21:51:05.538153 IP xyz.com.8104 > abc.com.8000: UDP, length 1316
21:51:05.538158 IP xyz.com.8104 > abc.com.8000: UDP, length 1316
21:51:05.538163 IP xyz.com.8104 > abc.com.8000: UDP, length 1316


So I am not exactly sure why it is not working over the cellular interface as 1316 is a reasonably low setting for MTU

I'll report back if I make more progress

**Reply by tmatth** [reply]
> @tmatth
> 
> I can confirm via tcpdump that the -pkt_size 1316 option is indeed working
> 
> 21:51:05.538147 IP xyz.com.8104 > abc.com.8000: UDP, length 1316 21:51:05.538153 IP xyz.com.8104 > abc.com.8000: UDP, length 1316 21:51:05.538158 IP xyz.com.8104 > abc.com.8000: UDP, length 1316 21:51:05.538163 IP xyz.com.8104 > abc.com.8000: UDP, length 1316
> 
> So I am not exactly sure why it is not working over the cellular interface as 1316 is a reasonably low setting for MTU
> 
> I'll report back if I make more progress

It's not uncommon to have to go as low as 1200, it depends what headers the network are adding like @aler9 was explaining.

**Comment by saket424** [comment]
Even lowering it to 1000 did not help with udp-based rtsp playback over cellular. Works over wifi just fine
Curiously tcp-based rtmp playback of the same feed over cellular works. Any ideas?

**Reply by tmatth** [reply]
rtsp-simple-server will reject UDP traffic if the source address of the incoming UDP traffic does not match the address that was signalled in the RTSP handshake, see:
https://github.com/aler9/rtsp-simple-server/issues/913
and
https://github.com/aler9/gortsplib/issues/107

So it's worth confirming that the IP address you see in tcpdump for the actual UDP packets matches the host in your RTSP signalling

**Comment by saket424** [comment]
@tmatth i see proper rtp packets flowing between the rtsp-simple-server and the android phone with the correct pkt_size+42 bytes to the t-mobile ip address 172.58.141.109 that did the signaling handshake. in addition to traffic to 172.58.141.109 i also see traffic from 172.58.139.232 which results in  destination port unreachable errors

Is there any was to temporarily loosen the UDP src address mismatch check to see if that  changes anything?

**Reply by tmatth** [reply]
You'd need to e.g., forward port this patch:
https://github.com/lionelnicolas/gortsplib/commit/ebbd6963cf321ee4c07af308a571fa224dae435a

[a comment by an automated account is not shown]
